### ISSUE TITLE
Benchmarking Fennel

### ISSUE THREAD

**Issue by shubham1172** [dburriss/fennel#3]
This is an awesome library! 

Are there any benchmarks available? Posting them here might help with production adoption as well. Thanks!

**Comment by dburriss** [comment]
Hey @shubham1172 
Thanks for the suggestion. I really just created this because I had a usecase and I thought I would share it. I also wanted an excuse to play with FParsec.
If I have a spare moment I may throw up a benchmark but I am not sure what I would benchmark it against? What I could do is list some through-put numbers for read and write, along with memory and CPU usage. 
I am not currently using this in any personal projects so it would be lower down in my list of OSS tickets to pick up.

**Comment by shubham1172** [comment]
@dburriss throughput numbers make sense!

**Comment by dburriss** [comment]
@shubham1172 I started playing around with some benchmarks. I may be doing something wrong but initially, this looks like it can parse about half a million metrics per second. 

| Method                | Mean             | Error            | StdDev           | Gen0        | Gen1       | Gen2      | Allocated    |
|---------------------- |-----------------:|-----------------:|-----------------:|------------:|-----------:|----------:|-------------:|
| Parse_1_Line          |         914.9 ns |          4.67 ns |          4.37 ns |      0.5274 |          - |         - |      1.08 KB |
| Parse_10_Lines        |      11,128.0 ns |         17.75 ns |         14.82 ns |      6.5308 |          - |         - |     13.35 KB |
| Parse_10_000_Lines    |  19,550,231.9 ns |    236,989.91 ns |    210,085.38 ns |   4812.5000 |  1843.7500 |  875.0000 |  13274.75 KB |
| Parse_500_000_Lines | 952,319,541.7 ns | 17,575,100.81 ns | 16,439,760.20 ns | 183000.0000 | 63000.0000 | 4000.0000 | 663686.77 KB |

**Comment by shubham1172** [comment]
This is really great and establishes confidence in using this library on hot paths, would you consider adding a `benchmark` folder and adding the code + results (with your machine specs)?

**Comment by dburriss** [comment]
Yep. I just need to clean it up a bit. I also measured the other way of creating string metrics.

| Method             | N     | Mean                | Error              | StdDev             | Gen0        | Gen1       | Gen2      | Allocated   |
|------------------- |------ |--------------------:|-------------------:|-------------------:|------------:|-----------:|----------:|------------:|
| Parse1     | N/A    |         877.9182 ns |          2.2461 ns |          2.1010 ns |      0.5274 |          - |         - |      1104 B |
| ParseBatchOf10     | 1     |      10,888.2554 ns |         22.7153 ns |         18.9683 ns |      6.5308 |          - |         - |     13672 B |
| CreateMetricString | 1     |         354.4184 ns |          1.8933 ns |          1.5810 ns |      0.3366 |          - |         - |       704 B |
| ParseBatchOf10     | 100   |   1,263,119.6669 ns |     11,570.9274 ns |     10,823.4527 ns |    542.9688 |   250.0000 |         - |   1359283 B |
| CreateMetricString | 100   |      35,205.3070 ns |        224.4907 ns |        199.0051 ns |     33.6304 |          - |         - |     70400 B |
| ParseBatchOf10     | 50000 | 898,562,752.8667 ns | 17,176,212.7793 ns | 16,066,640.0907 ns | 181000.0000 | 62000.0000 | 4000.0000 | 679613600 B |
| CreateMetricString | 50000 |  17,522,347.8371 ns |     58,958.5676 ns |     52,265.2345 ns |  16812.5000 |          - |         - |  35200030 B |
